include/aspect/newton.h

```diff
@@ -61,10 +61,28 @@ namespace aspect
    * A Class which can declare and parse parameters and creates
    * material model outputs for the Newton solver.
    */
+
   template <int dim>
   class NewtonHandler: public SimulatorAccess<dim>
   {
     public:
+
+      /**
+       * This enum describes the type of stabilization is used
+       * for the Newton solver. None represents no stabilization,
+       * SPD represent that the resulting matrix is make Symmetric
+       * Positive Definite, symmetric represents that the matrix is
+       * only symmetrized, and PD represents that we do the same as
+       * what we do for SPD, but without the symmetrization.
+       */
+      enum NewtonStabilization
```

Comment:
I had change this form `enum class NewtonStabilization` to enum NewtonStabilization` to make sure I could use the operator|. The downside of this is that is isn't strongly typed anymore. An other option I tried was to add a below this the function:

```
inline NewtonStabilization operator|(NewtonStabilization a, NewtonStabilization b)
{return static_cast<NewtonStabilization>(static_cast<int>(a) | static_cast<int>(b));}
```
but then I got the error:
```
error: ‘aspect::NewtonHandler<dim>::NewtonStabilization aspect::NewtonHandler<dim>::operator|(aspect::NewtonHandler<dim>::NewtonStabilization, aspect::NewtonHandler<dim>::NewtonStabilization)’ must take exactly one argument
       inline NewtonStabilization operator|(NewtonStabilization a, NewtonStabilization b)
```

Reply:
You can't make this function a member of the surrounding `NewtonHandler` class. It needs to be a function at namespace scope. Just declare it below the end of the `NewtonHandler` class as
```
inline
typename NewtonHandler<dim>::NewtonStabilization 
operator|(typename NewtonHandler<dim>::NewtonStabilization a, 
               typename NewtonHandler<dim>::NewtonStabilization b)
{
  return static_cast<typename NewtonHandler<dim>::NewtonStabilization>(
    static_cast<int>(a) | static_cast<int>(b));
}
```

Reply:
hmm, this somehow doesn't work. It doesn't seem to recognise that the inline opertor is pressent, but it compiles newton.h fine. So when I put:
```
  template <int dim>
  inline
  typename NewtonHandler<dim>::NewtonStabilization
  operator|(typename NewtonHandler<dim>::NewtonStabilization a,
                 typename NewtonHandler<dim>::NewtonStabilization b)
  {
    return static_cast<typename NewtonHandler<dim>::NewtonStabilization>(
      static_cast<int>(a) | static_cast<int>(b));
  }
```
I get as error:
```
error: no match for ‘operator|’ (operand types are ‘aspect::NewtonHandler<2>::NewtonStabilization’ and ‘aspect::NewtonHandler<2>::NewtonStabilization’)
               const double alpha = preconditioner_stabilization | NewtonHandler<dim>::NewtonStabilization::PD ?
```

Reply:
So I have tried to add `  NewtonHandler<dim>::NewtonStabilization operator|(typename NewtonHandler<dim>::NewtonStabilization a, typename NewtonHandler<dim>::NewtonStabilization b); \` to the bottom of newton.cc, but that doesn't help. I also tried to just explicity instanciate them:
```
  typename NewtonHandler<2>::NewtonStabilization
  operator|(typename NewtonHandler<2>::NewtonStabilization a,
                 typename NewtonHandler<2>::NewtonStabilization b)
  {
    return static_cast<typename NewtonHandler<2>::NewtonStabilization>(
      static_cast<int>(a) | static_cast<int>(b));
  }

  typename NewtonHandler<3>::NewtonStabilization
  operator|(typename NewtonHandler<3>::NewtonStabilization a,
                 typename NewtonHandler<3>::NewtonStabilization b)
  {
    return static_cast<typename NewtonHandler<3>::NewtonStabilization>(
      static_cast<int>(a) | static_cast<int>(b));
  }
```
This works, but then I will have to specifically convert it into an int or bool again because it complains: 
```
error: could not convert ‘aspect::operator|(velocity_block_stabilization, (aspect::NewtonHandler<2>::NewtonStabilization)1)’ from ‘aspect::NewtonHandler<2>::NewtonStabilization’ to ‘bool’
```
So if we really want this approach, we would also need to implement a way to covert this class to a bool.

Reply:
I think I understand why it doesn't work. 

Providing two specializations is ok with me. If you don't like this approach: What happens if you put the operator back into the `NewtonHandler` class as you had it initially, but prefix the whole declaration with the `friend` keyword? I.e., 
```
friend
NewtonStabilization 
operator| (const NewtonStabilization a, 
                const NewtonStabilization b)
{
  return static_cast<NewtonStabilization>(
    static_cast<int>(a) | static_cast<int>(b));
}
```
(But if you like the explicit specialization you show, then no need to try this approach.)

As for the conversion to `bool`: Just do the comparison via
```
  if ((value | ...::SPD) != ...::none)
```
`...::none` has numerical value zero, but has type equal to the `enum`, so you can compare with it.

Reply:
ok, I got it working now with the class enum. I do not have a strong preference for either way, but I guess it is nicer to not have explicit specializations.